### PR TITLE
[Tune] Catch SyncerCallback failure with dead node

### PR DESCRIPTION
**Pull request by justinvyu**
### Context
This issue was uncovered by this long running test: `long_running_distributed_pytorch_pbt_failure`. This test randomly kills nodes via `FailureInjectorCallback`, and the test failure happens when:

1. A trial result comes in and is processed
2. The node this trial is running on is requested to be killed by the failure injector
3. The driver's syncer callback runs on the on_trial_result event
4. The node dies
5. The driver is in the middle of syncing, trying to access the node ip, which errors

### What's in this PR?
1. Gracefully handle this race condition by catching the error thrown by the sync operation on a dead node
2. Log an error to the user
3. Adds a test for this sync with dead node scenario

## Related issue number

https://github.com/ray-project/ray/issues/29399

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by matthewdeng — APPROVED** [review]
Looks great, one minor comment is that we should not close the issue until this is cherry-picked into the release branch.

File: python/ray/tune/syncer.py [hunk 1]
```diff
@@ -545,7 +546,13 @@ def _sync_trial_dir(
         source_ip = self._trial_ips.get(trial.trial_id, None)
 
         if not source_ip:
-            source_ip = trial.get_runner_ip()
+            try:
+                source_ip = trial.get_runner_ip()
+            except RayActorError as e:
+                logger.error(
+                    f"Trial {trial}: An error occurred when trying to get the "
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply python/ray/tune/syncer.py:553]: I was following the other error messages below, and `Trial` has a string conversion already.

**Review by krfricke — APPROVED** [review]
Great! Shall we run the release test on this wheel?

**Comment by justinvyu** [comment]
> Great! Shall we run the release test on this wheel?

Started it!

**Pull request merged**